### PR TITLE
[WIP] Refactor dashboard_collection_link

### PR DESCRIPTION
**Pull request by josemigallas**
**What this PR does / why we need it**:

Proposal for refactoring `dashboard_collection_link` method. Reduces arguments in favor of using _options_ pattern.

**Which issue(s) this PR fixes** 

Comment from @hallelujah -> https://github.com/3scale/porta/pull/78#pullrequestreview-16212538

**Notes**
WIP⚠️ Waiting for vertical navigation branch to be merged

### CODE REVIEW

File: app/views/provider/admin/dashboards/_developers_navigation.html.slim [hunk 1]
```diff
@@ -4,17 +4,15 @@ nav.DashboardNavigation
...
-                                   'Accounts',
-                                   'users'
+                                   collection: buyers,
+                                   plural: 'Accounts',
+                                   icon: 'users'
 
       // Pending Accounts
       - if show_pending_accounts_on_dashboard?
         == dashboard_secondary_collection_link 'Pending',
-                                                pending_buyers,
```
[the remark this answers is not in the document]
Comment (thomasmaas) [reply app/views/provider/admin/dashboards/_developers_navigation.html.slim]: I had to give the method the plural in order to give it an icon

File: app/helpers/provider/admin/dashboards_helper.rb [hunk 2]
```diff
@@ -19,17 +19,22 @@ def sign_class(widget)
     widget.percentual_change > 0 ? 'u-plus' : 'u-minus'
   end
 
-  def dashboard_collection_link(singular_name, collection, path, plural = nil, icon_name = nil)
+  def dashboard_collection_link(singular_name, path, opts)
     link_to path, class: 'DashboardNavigation-link' do
-      link_text = pluralize(number_to_human(collection.size), singular_name, plural)
-      link_text = link_text.prepend "#{icon(icon_name)} " if icon_name
+      if opts[:collection]
+        link_text = pluralize(number_to_human(opts[:collection].size), singular_name, opts[:plural])
```
Comment (thomasmaas) [inline app/helpers/provider/admin/dashboards_helper.rb]: The idea of pluralise is that you give it the singular and it gives you the plural back. Not sure I follow this change.

Comment (josemigallas) [reply app/helpers/provider/admin/dashboards_helper.rb]: The purpose of the method wasn't clear but I realize now it is precisely to pluralize a link, isn't it? So I should not have made `collection` an optional argument.

File: app/views/provider/admin/dashboards/_developers_navigation.html.slim (hunk 1 above)
[the remark this answers is not in the document]
Comment (josemigallas) [reply app/views/provider/admin/dashboards/_developers_navigation.html.slim]: You're right, sorry

File: app/helpers/provider/admin/dashboards_helper.rb [hunk 3]
```diff
@@ -19,17 +19,17 @@ def sign_class(widget)
     widget.percentual_change > 0 ? 'u-plus' : 'u-minus'
   end
 
-  def dashboard_collection_link(singular_name, collection, path, plural = nil, icon_name = nil)
+  def dashboard_collection_link(singular_name, collection, path, opts)
```
Comment (hallelujah) [inline app/helpers/provider/admin/dashboards_helper.rb]: You can use keyword arguments

`def dashboard_collection_link(singular_name, collection, path, icon_name: nil, plural: nil)`

https://robots.thoughtbot.com/ruby-2-keyword-arguments

Comment (josemigallas) [reply app/helpers/provider/admin/dashboards_helper.rb]: Wasn't the point to reduce the method's number of arguments?

Comment (Martouta) [reply app/helpers/provider/admin/dashboards_helper.rb]: I don't think that this way is reducing the number of arguments but just hiding that we actually have more arguments 🤔 We would need an object then, but I still don't know how. I would have to look at how this is currently working 😄

Comment (josemigallas) [reply app/helpers/provider/admin/dashboards_helper.rb]: Creating an object is yet another way to hide arguments tho. I think the opts is a clean way to group optional arguments, reducing the signature of the method. But really, the whole point of this PR is to address @hallelujah 's comment -> https://github.com/3scale/porta/pull/78#discussion_r223078010 So.. I'm confused now.

Comment (Martouta) [reply app/helpers/provider/admin/dashboards_helper.rb]: His comment doesn't tell me that the goal is to have less params but that the goal is that the order to call to those params doesn't really matter (which is what you accomplish both with a hash and with what Hery suggested here).

About being the same with an object (so different methods), not really imo, for 2 reasons that I have in mind:
1. Being in a separate method means each method does independently less things.
2. As we have less possible paths for each method, and we have more than 1 method, it is easier to unit test them individually.
So doing it differently means changing the current logic so it is not hiding stuff, it is not the same. But it is clearly not easy to do either 😄

Comment (josemigallas) [reply app/helpers/provider/admin/dashboards_helper.rb]: You're right I misunderstood Hery's comment, so I'll remove `opts` and use the named arguments.
As for the object discussion, I don't this method is complex enough to justify using an object. I totally agree with you point, however for this particular case, it wouldn't make sense.

**Comment by Martouta** [comment]
Rebase from master plz 😄

**Comment by thomasmaas** [comment]
Lets hold off with merging until after we deployed vertical nav and all is working. Don't want to take any risk.

File: public/packs/manifest.json [hunk 4]
```diff
@@ -4,10 +4,10 @@
   "assets/glyphicons-halflings-regular.ttf": "/packs/assets/glyphicons-halflings-regular.e18bbf611f2a2e43afc071aa2f4e1512.ttf",
   "assets/glyphicons-halflings-regular.woff": "/packs/assets/glyphicons-halflings-regular.fa2772327f55d8198301fdb8bcfc8158.woff",
   "assets/glyphicons-halflings-regular.woff2": "/packs/assets/glyphicons-halflings-regular.448c34a56d699c29117adc64c43affeb.woff2",
-  "dashboard.js": "/packs/dashboard-bf5a87fcb060e76c9b61.js",
-  "dashboard.js.map": "/packs/dashboard-bf5a87fcb060e76c9b61.js.map",
-  "navigation.js": "/packs/navigation-c4c6613e2ce14fb63092.js",
-  "navigation.js.map": "/packs/navigation-c4c6613e2ce14fb63092.js.map",
-  "policies.js": "/packs/policies-221156de03fef790b50b.js",
-  "policies.js.map": "/packs/policies-221156de03fef790b50b.js.map"
+  "dashboard.js": "/packs/dashboard-1d1f099fe0e7e0ceba6b.js",
```
Comment (hallelujah) [inline public/packs/manifest.json]: Are we sure we are modifying the manifest?
@didierofrivia it looks like the packs were not saved correctly before

Comment (josemigallas) [reply public/packs/manifest.json]: I re-compiled Webpack and committed packs as well, waiting for @didierofrivia anyways.

File: app/helpers/provider/admin/dashboards_helper.rb [hunk 5]
```diff
@@ -19,17 +19,18 @@ def sign_class(widget)
     widget.percentual_change > 0 ? 'u-plus' : 'u-minus'
   end
 
-  def dashboard_collection_link(singular_name, collection, path, plural = nil, icon_name = nil)
+  def dashboard_collection_link(singular_name, collection, path, icon_name: nil, plural: nil, notice: false)
     link_to path, class: 'DashboardNavigation-link' do
       link_text = pluralize(number_to_human(collection.size), singular_name, plural)
       link_text = link_text.prepend "#{icon(icon_name)} " if icon_name
+      link_text = link_text.prepend ".u-notice" if notice
```
Comment (hallelujah) [inline app/helpers/provider/admin/dashboards_helper.rb]: This looks strange to me.

**Pull request closed without merging**